File: compiladores-2024-1/class1/linked_list.py
class Node:
    def __init__(self, data = None, next = None):
        self.data = data
        self.next = next

    def __repr__(self):
        return '%s -> %s' % (self.data, self.next)

class LinkedList:
    def __init__(self, head = None):
        self.head = head
    
    def insert_at_beginning(self, data):
        node = Node(data)
        node.next = self.head
        self.head = node

    def __repr__(self):
        return "[" + str(self.head) + "]"
    
    def insert_after(self, previous_node, data):
        node = Node(data)
        node.next = previous_node.next
        previous_node.next = node

    def find(self, data):
        actual_node = self.head
        while actual_node and actual_node.data != data:
            actual_node = actual_node.next
        
        return actual_node

    def remove(self, data):
        if self.head.data == data:
            self.head = self.head.next
        else:
            before_node = None
            actual_node = self.head

            while actual_node and actual_node.data != data:
                before_node = actual_node
                actual_node = actual_node.next
            
            if actual_node:
                before_node.next = actual_node.next
            else:
                before_node.next = None

# Testing the list:
list = LinkedList()
for i in range(8):
    list.insert_at_beginning(i)
print("List:", list)

for i in range(8, -4, -2):
    element = list.find(i)
    if element:
        print("List contains element {0}.".format(i))
    else:
        print("List doesn't contains element {0}.".format(i))

File: compiladores-2024-1/aula1/linkedList.go
package main

import "fmt"

type Node struct {
	data int
	next *Node
}
type LinkedList struct {
	head *Node
}

func (list *LinkedList) InsertAtBeginning(data int) {
	node := &Node{data: data}
	node.next = list.head
	list.head = node
}

func (list *LinkedList) print() {
	currentNode := list.head

	for currentNode != nil {
		fmt.Printf("%d -> ", currentNode.data)
		currentNode = currentNode.next
	}
}

func (list *LinkedList) remove(data int) {
	if list.head == nil {
		return
	}

	if list.head.data == data {
		list.head = list.head.next
		return
	}

	previousNode := list.head

	for previousNode.next != nil && previousNode.next.data != data {
		previousNode = previousNode.next
	}

	if previousNode.next == nil {
		fmt.Printf("List doesn't contains element.")
		return
	}

	previousNode.next = previousNode.next.next

}

func (list *LinkedList) find(data int) *Node {

	if list.head == nil {
		return nil
	}

	actualNode := list.head

	for actualNode != nil && actualNode.data != data {
		actualNode = actualNode.next
	}

	return actualNode
}
